lib/Dialect/Sim/SimOps.cpp

```diff
@@ -397,6 +397,223 @@ LogicalResult PrintFormattedProcOp::canonicalize(PrintFormattedProcOp op,
+    return failure();
+
+  bool isDeadOrEmpty = false;
+
+  auto *bodyBlock = &op.getBodyRegion().front();
+  isDeadOrEmpty = bodyBlock->without_terminator().empty();
+
+  if (!isDeadOrEmpty && !!op.getCondition())
+    if (auto cstCond = op.getCondition().getDefiningOp<hw::ConstantOp>())
+      isDeadOrEmpty = cstCond.getValue().isZero();
+
+  if (!isDeadOrEmpty)
+    return failure();
+
+  rewriter.eraseOp(op);
+  return success();
+}
+
+// --- TriggerSequenceOp ---
+
+LogicalResult TriggerSequenceOp::inferReturnTypes(
+    MLIRContext *context, std::optional<Location> location, ValueRange operands,
+    DictionaryAttr attributes, OpaqueProperties properties, RegionRange regions,
+    SmallVectorImpl<Type> &inferredReturnTypes) {
+  // Create N results matching the type of the parent trigger, where N is the
+  // specified length of the sequence.
+  auto lengthAttr =
+      properties.as<TriggerSequenceOp::Properties *>()->getLength();
+  uint32_t len = lengthAttr.getValue().getZExtValue();
+  Type trigType = operands.front().getType();
+  inferredReturnTypes.resize_for_overwrite(len);
+  for (size_t i = 0; i < len; ++i)
+    inferredReturnTypes[i] = trigType;
+  return success();
+}
+
+LogicalResult TriggerSequenceOp::verify() {
+  if (getLength() != getNumResults())
+    return emitOpError("specified length does not match number of results.");
+  return success();
+}
+
+LogicalResult TriggerSequenceOp::fold(FoldAdaptor adaptor,
+                                      SmallVectorImpl<OpFoldResult> &results) {
+  // Fold trivial sequences to the parent trigger.
+  if (getLength() == 1 && getResult(0) != getParent()) {
+    results.push_back(getParent());
+    return success();
+  }
+  return failure();
+}
+
+LogicalResult TriggerSequenceOp::canonicalize(TriggerSequenceOp op,
+                                              PatternRewriter &rewriter) {
+  if (op.getNumResults() == 0) {
+    rewriter.eraseOp(op);
+    return success();
+  }
+
+  // If the current op can be inlined into the parent,
+  // leave it to the parent's canonicalization.
+  if (auto parentSeq = op.getParent().getDefiningOp<TriggerSequenceOp>()) {
+    if (parentSeq == op) {
+      op.emitWarning("Recursive trigger sequence.");
+      return failure();
+    }
+    if (op.getParent().hasOneUse())
+      return failure();
+  }
+
+  auto getSingleSequenceUser = [](Value trigger) -> TriggerSequenceOp {
+    if (!trigger.hasOneUse())
+      return {};
+    return dyn_cast<TriggerSequenceOp>(trigger.use_begin()->getOwner());
+  };
+
+  // Check if there are unused results (which can be removed) or
+  // non-concurrent sub-sequences (which can be inlined).
+
+  bool canBeChanged = false;
+  for (auto res : op.getResults()) {
+    auto singleSeqUser = getSingleSequenceUser(res);
+    if (res.use_empty() || !!singleSeqUser) {
+      canBeChanged = true;
+      break;
+    }
+  }
+
+  if (!canBeChanged)
+    return failure();
+
+  // DFS for inlinable values.
```

Comment:
Can this be done instead on an op-by-op canonicalizer which doesn't walk the IR?  having DFS in a canonicalizer is a good way to have n^2 behavior.

Reply:
This is how I've implemented it originally, but I _think_ the current variant is more efficient. Note that:
- The first thing the canonicalizer does it to check, whether the current op can be inlined by the parent. If so, it bails out immediately.  That way we can guarantee that the DFS only is performed from the root of a collapsible sub-tree and the entire sub-tree is collapsed with a single op rewrite.
- The DFS will only enter a child node if it _will_ be inlined. Thus, it should not do any more traversal than an op-by-op canonicalizer would also need to do.
- The DFS itself is pretty cheap. Since all nodes are guaranteed to have a single predecessor we do not need to maintain a set of visited nodes to prevent walking into a cycle.

My motivation to replace the original implementation was to minimize the number of op rewrites. Since a TriggerSequenceOp can have a lot of results and users, I didn't want to substitute them repeatedly. The DFS makes sure this happens in one shot. Given all this, I do not see how this would cause a performance regression. But maybe I'm missing something?